### DISCUSSION TITLE
Zlib + c-ares  - can't build

### DISCUSSION THREAD

**Discussion by ghost** [curl/curl#8068] [Q&A] [answered]
Hello

I can't build zlib + cares for Windows VS 2019 in any way:

-Built with-ares x32, lib-debug(static):

> nmake /f makefile.msvc CFG=lib-debug 
 got the output file: libcaresd.lib

-Now I'm trying to build libcurl + c-ares (x32, debug, static) for the VS2019 command line, I enter the command:

> cd libcur_path/
> nmake /f Makefile.vc mode=static vc=16 debug=yes WITH_CARES=static CARES_PATH=c-ares_path/

Construction begins, but eventually ends with an error:

> libcaresd.lib(ares__addrinfo_localhost.obj) : error LNK2019: unresolved external symbol reference__imp__GetUnicastIpAddressTable@8 в функции _ares__system_loopback_addrs.
> libcaresd.lib(ares__addrinfo_localhost.obj) : error LNK2019: unresolved external symbol reference __imp__FreeMibTable@4 в функции _ares__system_loopback_addrs.
> ..\builds\libcurl-vc16-x86-debug-static-cares-static-ipv6-sspi-schannel\bin\curl.exe : fatal error LNK1120: unresolved external symbol reference: 2
> NMAKE : fatal error U1077: "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX86\x86\link.exe" : return code "0x460"
> Stop.
> NMAKE : fatal error U1077: "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX86\x86\nmake.exe" : return code "0x2"
> Stop.

I do not understand why, because I indicated the "statiс" everywhere.

**Comment by gvanem** [comment]
Try adding `iphlpai.lib` to `CARES_LIBS`.

**Reply by ghost** [reply]
I don't have this file and why should it be in the folder "CARES_LIBS" ?

**Reply by gvanem** [reply]
A typo; should be `iphlpapi.lib`. You'll have that.

**Reply by ghost** [reply]
Thank you, I found the file, but unfortunately adding the file to the folder "CARES_LIBS" - did not help. The error is the same.

**Comment by gvanem** [comment, marked as answer]
>  adding the file to the folder "CARES_LIBS" - did not help. 

Folder?? As I've told you already, make this patch:
```diff
--- a/winbuild/MakefileBuild.vc 2021-07-31 14:34:26
+++ b/winbuild/MakefileBuild.vc 2021-11-29 13:36:18
@@ -209,6 +209,8 @@
 CARES          = static
 !ENDIF

+CARES_LIBS     = $(CARES_LIBS) iphlpapi.lib
+
 !IFDEF USE_CARES
```

**Reply by ghost** [reply]
Sorry, now I understand.
Now everything works !!! Thank you!
Please tell me - this problem is related to this:
https://c-ares.org/mail/c-ares-archive-2009-05/0036.shtml

**Reply by gvanem** [reply]
This discussion should have been on the C-ares issues forum. It's kinda messy IMHO. Since:
* the functions `GetNetworkParams`, `GetAdaptersAddresses`, `GetBestRoute2` and `SystemFunction036` (!) are
  loaded dynamically from `iphlpapi.dll` (using `GetProcAddress()`).

* But `GetUnicastIpAddressTable` and `FreeMibTable` are not. Why?

And I do not know if the above email issue from 2009 is relevant now.
